[San Shao Qian] iP

SanShaoQian opened the pull request:
# StonksBot
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

StonksBot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103ay2425s2.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
> [!NOTE}
> StonksBot is a work in progress. Please be nice to it! 🥲

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SanShaoQian committed:
Rename chatbot

SanShaoQian committed:
Add greeting and exit

SanShaoQian committed:
Added Echo

SanShaoQian committed:
Add, List

Add the ability to store whatever text entered by the user and display them back to the user when requested.

SanShaoQian committed:
Added Mark as Done

Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

SanShaoQian committed:
Add ToDos, Events, Deadlines

Add support for tracking three types of tasks:

ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

SanShaoQian committed:
Add tests using the I/O redirection technique

Use the input/output redirection technique to semi-automate the testing of Duke.

SanShaoQian committed:
Add ability to handle errors

Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

SanShaoQian committed:
Add ability to delete tasks

Add support for deleting tasks from the list.

SanShaoQian committed:
Edit test and debug

SanShaoQian committed:
Add save function

Save the tasks in the hard disk automatically whenever the task list changes

SanShaoQian committed:
Add loading of data

Load the data from the hard disk when the chatbot starts up.

SanShaoQian committed:
Add more OOP

Refactor the code to extract out closely related code as classes.

Extract the following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list e.g., it has operations to add/delete tasks in the list

SanShaoQian committed:
Organise into packages

Organize the classes into suitable java packages.

SanShaoQian committed:
Add ability to process dates

Teach the chatbot how to understand dates and times.

SanShaoQian committed:
Merge branch 'add-gradle-support'

SanShaoQian committed:
Add JUnit tests

Added JUnit tests to test the behavior of Task, Todo and Parser

SanShaoQian committed:
Add JavaDoc comments

Added header comments to all non-private classes/methods, and non-trivial private methods.

SanShaoQian committed:
Edit code to coding standard

Tweak the code to comply with a given coding standard.

SanShaoQian committed:
Add a find function

Add functionality to find a task by searching for a keyword in the task description.

SanShaoQian committed:
Merge branch 'branch-A-CodingStandard'

SanShaoQian committed:
Merge branch 'branch-Level-9'

Xavierlhm commented on src/main/java/stonks/Stonks.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    public void run() {
+        ui.showGreeting();
+        boolean done = false;
```
Perhaps the naming could be changed to isDone to make it sound more like a Boolean Variable?

Xavierlhm commented on src/main/java/stonks/task/Task.java:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * Mark as undone
+     */
+    public void markNotDone() {
```
Good job following the coding standard for the naming of methods! 👍

Xavierlhm commented on src/main/java/stonks/ui/Ui.java:
```diff
@@ -0,0 +1,64 @@
+     * Display all tasks with description containing string provided
+     * @param taskManager contains all matching tasks
+     */
+    public void showMatchingList(TaskManager taskManager) {
```
Good job following the coding standard for the naming of methods! 👍

Xavierlhm reviewed (commented):
Good job having meaningful names for your methods! 👍

SanShaoQian committed:
Add a GUI

Used JavaFX to add a GUI to the chatbot

SanShaoQian committed:
Debug GUI

Integrate GUI with chatbot

lyhthaddeus commented on src/main/java/stonks/command/AddCommand.java:
```diff
@@ -0,0 +1,25 @@
+    public String execute(TaskManager tm, Storage storage) {
+        tm.addTask(task);
+        storage.save(tm.getTasks());
+        return "     Got it. I've added this task:\n"
```
could the 5 blank spaces in string be saved as a const?

lyhthaddeus commented on src/main/java/stonks/ui/Ui.java:
```diff
@@ -0,0 +1,58 @@
+     * Greets user when chatbot is opened
+     */
+    public void showGreeting() {
+        String greeting = "     Hello! I'm Stonks\n" +
```
Could all these spaces be saved as a const instead?
```suggestion
private String NEW_LINE_INDENT = "     ";
...
String greeting = NEW_LINE_INDENT + "Hello! I'm Stonks\n" + NEW_LINE_INDENT + "What can I do for you?";

lyhthaddeus commented on src/main/java/stonks/task/Task.java:
```diff
@@ -0,0 +1,45 @@
+ * An item to be completed
+ * Has a description and completion status
+ */
+public class Task {
```
could this be abstract  since it isnt a real object being used?

lyhthaddeus commented on src/main/java/stonks/parser/Parser.java:
```diff
@@ -0,0 +1,75 @@
+        case "unmark":
+            return new UnmarkCommand(Integer.parseInt(parts[1]) - 1);
+        case "todo":
+            String regex = "^\\s*(\\w+)\\s+(.*?)\\s*$";
```
I love the use of regex! :)

lyhthaddeus commented on src/main/java/stonks/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public String toString() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy");
```
Love the clean datetime formatting :)

lyhthaddeus reviewed (commented):
Overall your code looks great to read. Not many coding style violations (if any at all).
I really love the use of regex to help parse data (thats a really smart method).

jensenhuangyankai commented on src/main/java/stonks/storage/Storage.java:
```diff
@@ -0,0 +1,97 @@
+                boolean isDone = parts[1].equals("1");
+                String description = parts[2];
+                Task task = switch (type) {
+                    case "T" -> new Todo(description);
```
Good use of lambda switch case.

jensenhuangyankai commented on src/main/java/stonks/ui/Ui.java:
```diff
@@ -0,0 +1,58 @@
+     * @param message text to be displayed
+     */
+    public void showMessage(String message) {
+        String SEPARATOR = "    ____________________________________________________________\n";
```
Good that your constand was named using a Screaming snake case. However, consider putting final keyword on it

jensenhuangyankai commented on src/main/java/stonks/task/TaskManager.java:
```diff
@@ -0,0 +1,83 @@
+     * Returns total number of tasks
+     * @return size of list of tasks
+     */
+    public int size() {
```
should use a verb for this, like getSize()

SanShaoQian committed:
Add assertions

Use assert feature to document important assumptions that should hold at various points in the code

SanShaoQian committed:
Improve Code Quality

Refactor the code to improve the code quality

SanShaoQian committed:
Merge pull request #1 from SanShaoQian/branch-A-Assertions

Add assertions

SanShaoQian committed:
Merge pull request #2 from SanShaoQian/branch-A-CodeQuality

Improve Code Quality

SanShaoQian committed:
Add duplicate detection

addTask(task) method adds all tasks it is provided

there can be duplicate events, or existing events need to be edited

Check all added tasks to see if there are common descriptions. If yes, remove prior task and update new task instead, preserving isDone status

Ensure consistency of todo list by adding to same location

SanShaoQian committed:
Merge pull request #3 from SanShaoQian/branch-BCD-Extension

Add duplicate detection

realidx reviewed (commented):
The code quality is good

SanShaoQian committed:
Fix minor bugs

SanShaoQian committed:
Update README.md

Add a User Guide to the project

SanShaoQian committed:
Delete unused files

SanShaoQian committed:
Fix minor bugs
